[ao][sparsity] Basic implementation of activation sparsifier

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81814
* #80890
* #80889
* #80888
* #80887
* __->__ #80886

The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

Commit:
[ao][sparsity] Basic implementation of activation sparsifier

The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

z-a-f commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -0,0 +1,96 @@
+        return out
+
+
+class _ActivationSparsifierUtils:
```
It looks like this class is just a collection of functions to test. Can we remove it or pass the dict of functions directly? Not sure you need a factory here

macandro96 replied:
The idea was to organize the configs into a separate class for better readability purposes. Moved it to main Test Class.

z-a-f commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -0,0 +1,96 @@
+        }, sparse_config
+
+
+class _ActivationSparsiferTestCase(TestCase):
```
Is there a reason this is a separate test class? It looks like it is being used only in the TestActivationSparsifier -- why not just using that?

z-a-f commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -0,0 +1,96 @@
+        # simulate the workflow
+
+        # STEP 1: make data and activation sparsifier object
+        defaults, sparse_config = _ActivationSparsifierUtils._get_defaults()
```
There is no point for the container class -- just pass them directly. That would improve the readability of this file

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -0,0 +1,149 @@
+    Note::
+        The sparsification mask is computed on the input **before it goes through the attached layer**.
+
+    Args:
```
Please, follow the sphinx syntax: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

- No blanks between args
- Args should end with a colon

macandro96 replied:
done

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -0,0 +1,149 @@
+            The model whose layers will be sparsified. The layers that needs to be
+            sparsified should be added separately using the register_layer() function
+
+        aggregate_fn (Optional, function type)
+            default aggregate_fn that is used if not specified while registering the layer.
+            specifies how inputs should be aggreagated over time.
+
+        reduce_fn (Optional, function type)
+            default reduce_fn that is used if not specified while registering the layer.
+            reduce_fn will be applied on the aggregated input
+
+        mask_fn (Optional, function type)
+            default mask_fn that is used to create the sparsification mask (based on the
+            aggregated & reduced input). This is used if not specified while registering the layer
```
`function type` is often referred to as `Callable`

z-a-f replied:
For all callables it is usually a good idea to include information how that callable will be used / called. That would give information to the developers on what arguments should a function expect.

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -0,0 +1,149 @@
+            Default configuration for the mask_fn. This config will be passed
+            with the mask_fn()
+
+    Expected Usage:
```
nit: Replace with `Example:`

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -0,0 +1,149 @@
+    Expected Usage:
+        >>> model = SomeModel()
+        >>> act_sparsifier = ActivationSparsifier(...)  # init activation sparsifier
+
```
nit: Avoid blank spaces within the code (unless intended)

HDCharles commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -0,0 +1,149 @@
+            default aggregate_fn that is used if not specified while registering the layer.
+            specifies how inputs should be aggreagated over time.
+
+        reduce_fn (Optional, function type)
```
its hard to understand what the aggregate and reduce_fn do, also if there are defaults, shouldn't those be specified?

HDCharles commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -0,0 +1,149 @@
+
+        >>> sparsifier.squash_mask()
+    """
+    def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=None,
```
should this be passed in more along the lines of the normal sparsifier where a sparse defaults object is passed in?

macandro96 replied:
So, there are two kinds of config - 
1. function and feature level config: includes aggregate_fn, reduce_fn, mask_fn, features, etc. 
2. sparse config: Sparsity based config. These arguments are passed along with the mask_fn()

The normal sparsifier accepts sparse_config only where as in here we ask user to specify function, feature level config (explicit args) and sparse_config (similar to normal sparsifier).

HDCharles reviewed (approved):
will approve but I feel the API should be brought in line with the normal sparsifier unless there's a good reason.

jerryzh168 commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -0,0 +1,96 @@
+        combined_defaults = {**defaults, 'sparse_config': sparse_config}
+
+        # check model
+        assert activation_sparsifier.model == model
```
feels like we are checking implementation details here, I feel we do not need to do that

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Basic implementation of activation sparsifier"


The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
